ably/sharded_task.go

```diff
@@ -0,0 +1,159 @@
+)
+
+func generateChannelName(testConfig TestConfig, number int) string {
+	return "sharded-test-channel-" + strconv.Itoa(number%testConfig.NumChannels)
+}
+
+func randomChannelName(testConfig TestConfig) string {
+	r := rand.Intn(testConfig.NumChannels)
+
+	return generateChannelName(testConfig, r)
+}
+
+func publishOnInterval(testConfig TestConfig, ctx context.Context, channel *ably.RealtimeChannel, delay int) {
+	log.Println("Delaying publish to", channel.Name, "for", delay+testConfig.PublishInterval, "seconds")
+	time.Sleep(time.Duration(delay) * time.Second)
+
+	ticker := time.NewTicker(time.Duration(testConfig.PublishInterval) * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			log.Println("Publishing to:", channel.Name)
+
+			data := randomString(testConfig.MessageDataLength)
+			timePublished := strconv.FormatInt(millisecondTimestamp(), 10)
+
+			_, err := channel.Publish(timePublished, data)
+
+			if err != nil {
+				boomer.RecordFailure("ably", "publish", 0, err.Error())
+			} else {
+				boomer.RecordSuccess("ably", "publish", 0, 0)
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func shardedPublisherTask(testConfig TestConfig) {
+	ctx, cancel := context.WithCancel(context.Background())
+	boomer.Events.Subscribe("boomer:stop", cancel)
+
+	for i := 0; i < testConfig.NumChannels; i++ {
+		channelName := generateChannelName(testConfig, i)
+
+		client, err := newAblyClient(testConfig)
+		if err != nil {
+			boomer.RecordFailure("ably", "publish", 0, err.Error())
+			return
+		}
+		defer client.Close()
+
+		channel := client.Channels.Get(channelName)
+		defer channel.Close()
+
+		delay := i % testConfig.PublishInterval
+
+		go publishOnInterval(testConfig, ctx, channel, delay)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func shardedSubscriberTask(testConfig TestConfig) {
+	ctx, cancel := context.WithCancel(context.Background())
+	boomer.Events.Subscribe("boomer:stop", cancel)
+
+	channelName := randomChannelName(testConfig)
+
+	log.Println("Subscribing to channel:", channelName)
+
+	client, err := newAblyClient(testConfig)
+
+	if err != nil {
+		boomer.RecordFailure("ably", "subscribe", 0, err.Error())
+		return
+	}
+	defer client.Close()
+
+	channel := client.Channels.Get(channelName)
+	defer channel.Close()
+
+	sub, err := channel.Subscribe()
+	if err != nil {
+		boomer.RecordFailure("ably", "subscribe", 0, err.Error())
+		return
+	}
+	defer sub.Close()
+
+	connectionStateChannel := make(chan ably.State)
+	client.Connection.On(connectionStateChannel)
+
+	var lastDisconnectTime int64 = 0
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case connState := <-connectionStateChannel:
+			if connState.State == ably.StateConnDisconnected {
+				lastDisconnectTime = millisecondTimestamp()
+			} else if connState.State == ably.StateConnConnected && lastDisconnectTime != 0 {
+				timeDisconnected := millisecondTimestamp() - lastDisconnectTime
```

Comment:
I'm assuming that there is only one connection? You might want to reset the `lastDisconnectTime` here. Checking the `lastDisconnectTime != 0` implies that you're not confident that you will receive a corresponding disconnect.

Reply:
When a connection is first established, there isn't a corresponding disconnect.